jerry-core/ecma/base/ecma-helpers-conversion.c

```diff
@@ -457,6 +484,57 @@ ecma_utf8_string_to_number (const lit_utf8_byte_t *str_p, /**< utf-8 string */
     return ecma_number_make_infinity (sign);
   }
 
+  if (!is_infinity_check)
+  {
+    lit_utf8_size_t e_position = 0;
+    lit_utf8_size_t dot_position = 0;
+
+    while (e_position < str_size)
+    {
+      if (str_p[e_position] == LIT_CHAR_DOT)
+      {
+        dot_position = e_position;
+      }
+      else if (str_p[e_position] == LIT_CHAR_LOWERCASE_E || str_p[e_position] == LIT_CHAR_UPPERCASE_E)
+      {
+        break;
+      }
+
+      e_position++;
+    }
+
+    if (e_position != str_size && e_position != 0)
+    {
+      ecma_number_t base = ecma_utf8_string_to_number (str_p, e_position, true);
+      ecma_number_t exponent = ecma_utf8_string_to_number (str_p + e_position + 1, str_size - e_position - 1, true);
+      base /= pow (10, e_position - dot_position - 1);
+
+      /* Normalizing base */
+      while (base >= 10.0)
+      {
+        base /= 10.0;
+        exponent++;
+      }
+
+      /** If the convertable number is bigger than Number.MAX_VALUE
+        * then return with +- Infinity */
+      if (str_p[e_position + 1] != LIT_CHAR_MINUS &&
+          (exponent > NUMBER_MAX_VALUE_EXPONENT ||
+           (exponent == NUMBER_MAX_VALUE_EXPONENT && base > NUMBER_MAX_VALUE_BASE)))
+      {
+        return ecma_number_make_infinity (sign);
+      }
```

Comment:
Aren't this `if` and the one below, which deal with the exponent (sometimes together with the "base") being too big or too small, the core of this fix? Couldn't these be moved to the exponent parsing part where `e`, `e_sign`, `fraction_uint64 ` are already known (or where they are being computed), and we could make the same decisions without any new arguments and recursion? (I have a very strong gut feeling that it has to be solvable that way.)

Reply:
I see your point and with a quick research I may found a solvable solution for it!